rfcs/2021-07-20-8288-csv-enrichment.md

```diff
@@ -0,0 +1,170 @@
+# RFC 8288 - 2021-07-20 - CSV Enrichment
+
+Vector needs to allow users to enrich the events flowing through the topology
+using through the topology using a CSV file. This RFC proposes adding a Vrl
+function that can perform this lookup.
+
+## Scope
+
+The RFC will cover a simple VRL function that will lookup a single row from a
+CSV file. 
+
+The lookup will be a simple sequential scan. The first version of this
+function will not handle any indexing of this data. Above a certain size
+indexing will provide substantial performance benefits, so it is something that
+we should definitely plan for implementing at a future stage. The RFC will 
+briefly discuss this, but will not go into any specific indexing strategies.
+
+## Motivation
+
+Users would like to enrich events flowing through Vector with extra data 
+provided by a CSV file. 
+
+## Doc-level Proposal
+
+Two remap functions:
+
+### `lookup_csv_row`
+
+This function will look up a single row within the dataset. If a single row is 
+found that data is returned as an object, otherwise this function will error. 
+
+### `lookup_csv_rows` 
+
+This function returns the results as an array of objects. If no row is found an
+empty array is returned. Multiple results just result in multiple rows in the 
+array.
+
+### Parameters 
+
+*filename*
+
+The filename is the path to the filename of the csv file to lookup. This
+filename has to be a static string (cannot be changed at runtime) so that 
+Vector can:
+
+1. Validate that the file exists.
+2. Load the file into memory.
+
+The filename will be relative to the current working directory that Vector is
+running in.
```

Comment:
So, I'm not convinced this is the right strategy to take.

My preference would be as follows:

1. Allow to define a list of paths to which the VRL compiler/runtime has access in Vector's TOML file
2. Pass those to the VRL compiler and add them to the VRL context object
3. Introduce a new `p'...'` static type that represents a relative file location (i.e. it's a compile-time error to provide an absolute path or an invalid path string), and matches the first file it finds within the provided list of root paths defined in the Vector config.

Here's an example:

```toml
[transforms.remap]
type = "remap"
whitelist_paths = ["/opt/mycsvs"]
source = """
	lookup_csv_row(.data, p'my.csv')
"""
```


This has the following advantages:

1. It separates path location access from VRL source, meaning your VRL source easier to maintain (especially when we introduce loading VRL from a separate file (#6403).
2. It allows us to provide very specific compiler errors for path errors (e.g. invalid path syntax, non-existing files, read-access, etc)
3. Eventually (maybe, if need be), we can lift the `whitelist_paths` (or whatever we'll name it) up into a top-level Vector configuration option for other components to use.

Reply:
I'm wondering if it might be even better to move the paths *and* files out of the transform altogether.  @zsherman had a sample config example along the lines of:

```
global.csvfile = "/data.csv"

[transforms.enrich]
  source = '''
     . |= lookup_csv_row(global.csvfile, { "name": .name })
  '''
```

The advantage of this, in addition to the benefits you have mentioned, it also allows transforms to share the one data file. It's certainly possible that a given configuration may want multiple different transforms all enriching from the same datafile.